---
title: "Repguide vignette"
author: "David Brocks"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Repguide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

# Workflow
In this tutorial, we are going to design guideRNAs against long terminal repeats (LTRs) originating from the human endogenous retrovirus 9. By following this workflow, you will understand the basic steps of the Repguide package and learn how to adapt them to your custom requirements.

## Getting started

### Genome assembly
Repguide requires the genome assembly of your organism of choice stored as a _BSgenome_ object. You can retrieve the full list of supported genomes by typing `BSgenome::available.genomes()` or create a custom _BSgenome_ object following the [instructions](https://www.bioconductor.org/packages//2.7/bioc/vignettes/BSgenome/inst/doc/BSgenomeForge.pdf). 

Since we are designing guideRNAs against LTRs of the human genome, we will first install the UCSC hg38 assembly.
```{r, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
```
__Important__ __note__: Do not use repeat-masked BSgenome objects (contain 'masked' suffix, e.g. _BSgenome.Hsapiens.UCSC.hg38.masked_)!

### Genome annotation
In addition to the assembled genome sequence, Repguide imports genomic coordinates of transposable elements (mandatory) and cis regulatory features (optional). A convient solution is to download transposon coordinates using the [UCSC table browser](https://genome.ucsc.edu/cgi-bin/hgTables) by selecting your genome of choice, group = Repeats, track = RepeatMasker, and output format = all fields from selected table. You can provide custom annotation as long as it provides the basic information about chromosome, start, end, strand, repname (family identifier), and te_id (unique locus identifier). In this tutorial, we will show you how to import such information using the provided example datasets. 

### Bowtie aligner
Repguide uses a R implementation of the short read aligner [Bowtie](http://bowtie-bio.sourceforge.net/index.shtml) to map potential guideRNAs against the genome. Bowtie requires a FM index based on the Burrows-Wheeler transform to do so. 
In case Repguide does not find the corresponding bowtie index files to your chosen genome, it will automatically create them. Since this take about an hour (depends on workstation and genome), you can also [download](ftp://ftp.ccb.jhu.edu/pub/data/bowtie_indexes/) the appropriate index files (if available), rename them following the format (BSgenome + suffix, see below) and provide the directory in the `refdir` argument of the `createGuideSet()` function.

Example naming of the human hg38 bowtie indeces:  

 * BSgenome.Hsapiens.UCSC.hg38.1.ebwt  
 * BSgenome.Hsapiens.UCSC.hg38.2.ebwt  
 * BSgenome.Hsapiens.UCSC.hg38.3.ebwt (not required)  
 * BSgenome.Hsapiens.UCSC.hg38.4.ebwt (not required)  
 * BSgenome.Hsapiens.UCSC.hg38.rev.1.ebwt  
 * BSgenome.Hsapiens.UCSC.hg38.rev.2.ebwt  

## Creating a guideSet

We start the workflow by loading Repguide and the human hg38 _BSgenome_ object into our R environment:
```{r, warning = FALSE, message = FALSE}
library(Repguide)
library(BSgenome.Hsapiens.UCSC.hg38)
```
 
We then create a _guideSet_ object, which is a S4 class to store genomic information and results of later functions. To create such object, we have to define the genome assembly and provide annotation on transposable elements and cis regulatory features (e.g. promoters/enhancers). While the latter is optional, we highly recommend it for optimal guideRNA design. In this tutorial, we will use the hg38 [UCSC](https://genome.ucsc.edu/cgi-bin/hgTables) repeatmasker output and [FANTOM5 TSS annotation](http://fantom.gsc.riken.jp/5/). We additionally blacklist promoter regions of essential human genes (see `?essential_genes` for further information). Those three datasets are included in the Repguide package. We can also specify the path to the corresponding bowtie index directory using the `refdir` parameter. If `refdir = NULL` (the default), Repguide will search for bowtie indeces in the package installation directory and create the indeces there if not found.
```{r, warning = FALSE}
# Path to FANTOM5 hg38 TSS bed file (provided)
tss_path <- system.file(package = 'Repguide', 
                        'extdata', 
                        'hg38_fair+new_CAGE_peaks_phase1and2.bed.gz')
# Path to RepeatMasker hg38 repeat annotation (provided, LTRs only)
repeats_path <- system.file(package = 'Repguide', 
                            'extdata', 
                            'hg38_ucsc_rmsk_ltr.txt.gz')
# Path to essential genes (provided)
essentials_path <- system.file(package = 'Repguide', 
                               'extdata', 
                               'hg38_essential_gene_tss_coords.bed')

gs <- createGuideSet(genome = Hsapiens,           # Hsapiens is the short version of BSgenome.Hsapiens.UCSC.hg38
                     tes = repeats_path,          # TE annotation
                     cis = tss_path,              # TSS and/or enhancer annotation 
                     blacklist = essentials_path) # Regions to blacklist
gs                     
```

## Adding target loci
The generated _guideSet_ contains all annotation required to design guideRNAs against LTRs of the human genome. To add target sites, we can either provide names of repeat families we wish to target or a GRanges object with coordinates of the loci (__Important__ __note__: Such a GRanges object requires repname and te_id metadata columns compatible with the _guideSet's_ TE annotation).

We can retrieve the full list of available repeat families using the `repnames()` accessor function or search for specific families that match a specific pattern. For example, LTRs originating from the human endogenous retrovirus 9 all conveniently share a common 'LTR12' prefix (The [_Dfam_]( http://dfam.org/entry/DF0000399/relationships) database allows open access to transposon genetic relationships), allowing us to easily retrieve all its members by using regular expression:
```{r}
families <- repnames(gs, pattern = 'LTR12')
families
```

We now add these families as targets to our previously generated _guideSet_:
```{r, warning = FALSE, message = FALSE}
gs <- addTargets(gs, 
                 targets = families) # Either character vector of repname identifier or GRanges object
```

To inspect targeted families more closely or to design guideRNAs against specific parts of the families' consensus model, we can compute or import multiple sequence alignments. Repguide computes a rough multiple sequence alignment by weighted sampling from a families' total genomic sequences. While this approach is fast and usually sufficient to generate a representative consensus model, it might fail for highly polymorphic families. Therefore, Repguide also supports reading pre-computed multiple sequence alignments stored in fasta format. To import several such alignments, we first need to create a data.frame containing the family identifier of the alignment and the corresponding file's location:
```{r, warning = FALSE}
# Path to data directory containg multiple sequence alignment files
data_dir <- system.file(package = 'Repguide', 'extdata')

# Return paths to files ending with *.fasta.gz in that data directory
msa_files <- dir(data_dir, full = TRUE, pattern = 'fasta.gz$')

# Create data.frame with repname and corresponding file path
msa_df <- data.frame('repname' = c('LTR12B', 'LTR12D'), 'path' = msa_files)
msa_df

gs <- addAlignments(gs,                 # our guideSet
                    files = msa_df,     # our data.frame
                    max_gap_freq = 0.8) # Removes columns with more than 80% gaps in the alignment
```

Having added the targets and alignments to our _guideSet_, we can now generate several QC reports and inspect the targeted families more closely:
```{r, fig.width = 7, fig.height=7, fig.align='center', warning = FALSE, message = FALSE}
gs <- plotTargets(gs) # stores QC plots in guideSet and prints them to screen
```

The output generates general statistics on selected families and optional information that requires a multiple sequence alignment (see `?plotTargets` for full annotation of the panels). As we can see from the figures, members of the LTR12C family are by far most prevalent and largest in size (panels __a__ to __c__). Since we also calculated and added alignments, the output further visualizes the multiple sequence alignment (__d__), the density of CpG dinucleotides across it (__e__) and includes information on putative consensus transcription start sites (TSSs) (__f__). We can use this information for example, to optimize the design of our guideRNAs by targeting CpG dense regions close to the putative TSS.

## Adding guideRNAs
After we are satisfied with the loci/families we wish to target, we continue to compute all possible guideRNAs fulfilling any user-defined criteria (e.g. GC content, target/off-target score, or binding position on consensus models). In this tutorial, we will restrict ourselves to the design of 19 bp guideRNAs with a 3' NGG PAM. We further tell Repguide to score guides based on their binding to CpG dense parts of the LTR12B and LTR12E families. 

For that purpose, we have to first create a simple data.frame providing coordinates on the consensus we wish to target. The `plotTargets()` report can assist you in deciding what start and end coordinates to supply.
```{r}
consensus_df <- data.frame('repname' = c('LTR12B', 'LTR12E'), 
                           'start' = c(1, 1), 
                           'end' = c(150, 500))
consensus_df                           
```
_Note_ _that_ _guide_ _target_ _binding_ _sites_ _outside_ _the_ _consensus_ _range_ _will_ _be_ _scored_ _neutrally_.  

We then call the `addGuides()` function to compute our guide universe. More specifically, this function first calls [_jellyfish_](https://www.cbcb.umd.edu/software/jellyfish/) to calculate the 19mer (in our scenario) universe of specified targets, selects kmers with proper PAM sequence, and finds and annotates their putative genomic binding sites. Since most kmers are genetically similar and hence show redundanct binding profiles, Repguide finally clusters kmers into groups with homogenous binding profiles and marks the best kmer per cluster (the number of groups and selection procedure can be specified with the `n_clust` and `alpha` arguments of the `addGuides()` function). Depending on the target sequence complexity, guideRNA length, and allowed mismatches, the computation may take a bit.

```{r, warning = FALSE, message = TRUE}
gs <- addGuides(gs,                             # our guideSet
                n_mismatches = 2,               # max allowed mismatch of reported binding sites
                min_Son = 50,                   # minimal score requirement for valid guides
                guide_length = 19,              # length of the guideRNAs to design
                consensus_range = consensus_df, # restrict guideRNA design to parts on consensus
                gc_content = c(0.4, 0.8))       # allowed guide GC content (between 40 and 80 percent)
```

We can now generate a QC report and sanity check the results by entering:
```{r, fig.width = 7, fig.height=12, fig.align='center', warning = FALSE, message = FALSE}
gs <- plotGuides(gs)
```

We can see from the report that while most of the potential guides have the desired GC content, only a minority passes our minimal score requirements (__a__ - __c__). Panel __d__ shows the Sbind score distribution by the total number of mismatches of a guide's genomic complement and __e__ summarizes the frequency of off-targets by their basepair distance to the nearest cis regulatory element. The dotted red line indicates the conversion of basepair distances to the corresponding cis score (Scis) of a binding site. Panels __f__ and __g__ show the clustering of guideRNAs by their target binding profile and selection of the best guide per cluster, respectively. For full annotation of the panels, we refer to the documentation page of the `plotGuides()` function.

To adjust the guide selection or clustering steps, you can use the `selGuides()` or `clustGuides()` functions, respectively, without re-running the entire `addGuides()` pipeline.

## Calculating guide combinations
Multiplexing of several distinct guideRNAs for combined delivery may maximize recruitment of dCAS9 to polymorphic multi-copy families, as recently shown for human LTR5HS elements [1]. However, selecting the optimal set of guideRNAs for maximal on- and minimal off-targeting is not trivial. Repguide first computes the binding profiles of all possible combinations of previously selected guide representatives, i.e. the best guides per cluster. The best combination then serves as seed to initialize a greedy algorithm (can be disabled by setting `greedy = FALSE`) to further optimize the set. The `alpha` parameter controls the weight of off-targets, i.e. `alpha = 0` maximizes target coverage regardless of off-binding score while larger `alpha` penalizes unspecific sets of guideRNAs (see [guideRNA scoring](https://tanaylab.bitbucket.io/Repguide/articles/guide_scoring.html) for further information).

We calculate binding profiles of all possible combinations for up to 6 distinct guides by typing:
```{r}
gs <- addCombinations(gs,             # our guideSet
                      iterations = 10, # number of greedy search iterations
                      greedy = TRUE,  # run greedy algorithm
                      alpha = 100,    # off-target score coefficient
                      max_guides = 6) # maximal number of guides to consider
```

Having added the score statistics for computed combinations to our _guideSet_, we can inspect the results by entering:
```{r, fig.width = 7, fig.height = 12, fig.align = 'center', warning = FALSE, message = FALSE}
gs <- plotCombinations(gs)
```

The table in panel __a__ summarizes the on- and off-targets per guide combination. We can see that the on-target binding score plateaus with increasing number of distinct guides (__a__ and __b__), meaning the addition of further guideRNAs to the delivery construct would provide little extra benefit. Panel __c__ visualizes the performance of the greedy algorithm and panel __d__ shows on- vs off-target scores for all computed guide combinations. Closer inspection of predicted off-targets is shown in panels __e__ and __f__ which summarize off-target binding coverage by genomic feature and distance to nearest cis regulatory element, respectively. The target loci coverage per number of guides is shown in panel (__g__). Note how several loci are covered by more than one binding site. Panel __h__ shows the predicted binding position along the consensus sequences (in 250 bp bins) by the number of guides. Full annotation of the figure is available in the documentation of the `plotCombinations()` function.

## Exporting the results
We can finally save the relevant results and plots of our guideSet to a specified directory:
```{r, eval = FALSE}
export(gs,
       outdir = tempdir(), 
       dpi = 320, # image resolution,
       full = TRUE, # exports stats for all guides, not only selected ones
       workspace = FALSE) # don't save guideSet as workspace
```

Executing the command will create a subdirectory in the specified outdir (or current working directory if `force = TRUE`) and export the results.
```{r, eval = FALSE}
knitr::include_graphics(img1_path)
```   

# References
[1] Daniel R Fuentes, Tomek Swigut, Joanna Wysocka. Systematic perturbation of retroviral LTRs reveals widespread long-range effects on human gene regulation. eLife (2018).  

# Session information
```{r}
sessionInfo()
```          