<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Benchmarking â€¢ Repguide</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Benchmarking">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Repguide</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="../articles/Repguide.html">
    <span class="fa fa-play-circle"></span>
     
    Get started
  </a>
</li>
<li>
  <a href="../articles/benchmark.html">
    <span class="fa fa-bar-chart"></span>
     
    Benchmark
  </a>
</li>
<li>
  <a href="../articles/guide_scoring.html">
    <span class="fa fa-star"></span>
     
    Guide scoring
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-th-list"></span>
     
    Reference
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-news"></span>
     
    News
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/tanaylab/repguide">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Benchmarking</h1>
                        <h4 class="author">David Brocks</h4>
            
            <h4 class="date">2019-09-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/tanaylab/repguide/blob/master/vignettes/benchmark.Rmd"><code>vignettes/benchmark.Rmd</code></a></small>
      <div class="hidden name"><code>benchmark.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>In this benchmark, we will use Repguide to predict genomic binding sites for a publically available set of 12 guideRNAs targeting LTR5_Hs long terminal repeats and compare the results to experimentally determined Cas9 ChIP-seq peak intensities [1].</p>
</div>
<div id="running-repguide" class="section level2">
<h2 class="hasAnchor">
<a href="#running-repguide" class="anchor"></a>Running Repguide</h2>
<p>We first create a hg38 guideSet and provide the guide sequences to map.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(Repguide)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BSgenome.Hsapiens.UCSC.hg38)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyverse)

<span class="co"># Path to RepeatMasker hg38 repeat annotation (provided, LTRs only)</span>
repeats_path &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="dt">package =</span> <span class="st">'Repguide'</span>, <span class="st">'extdata'</span>, <span class="st">'hg38_ucsc_rmsk_ltr.txt.gz'</span>)

gs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/createGuideSet.html">createGuideSet</a></span>(<span class="dt">genome =</span> Hsapiens, <span class="co"># Hsapiens is the short version of BSgenome.Hsapiens.UCSC.hg38</span>
                     <span class="dt">tes =</span> repeats_path) <span class="co"># can be either path to file or GRanges object)</span>
<span class="co">#&gt; Created new guideSet at Mon Sep 16 10:06:29 2019</span>
                     
gs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addTargets.html">addTargets</a></span>(gs, <span class="st">'LTR5_Hs'</span>)

<span class="co"># Import guideRNA sequences used in Fuentes et al. [1] against LTR5_Hs</span>
guides_path &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="dt">package =</span> <span class="st">'Repguide'</span>, <span class="st">'extdata'</span>, <span class="st">'fuentes_elife_2018_guides.fasta'</span>)
guides_fuentes &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(Biostrings<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span>(guides_path))

gs_fuentes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addGuides.html">addGuides</a></span>(gs, 
                        <span class="dt">guides =</span> guides_fuentes, 
                        <span class="dt">n_mismatches =</span> <span class="dv">3</span>, 
                        <span class="dt">min_Son =</span> <span class="dv">0</span>, 
                        <span class="dt">max_Soff =</span> <span class="ot">Inf</span>, 
                        <span class="dt">gc_content =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>,<span class="dv">1</span>),
                        <span class="dt">PAM =</span> <span class="st">'NGG'</span>)                     
<span class="co">#&gt; [1] "Aligning kmers against: BSgenome.Hsapiens.UCSC.hg38"</span>
<span class="co">#&gt; [1] "Annotating guides"</span>
<span class="co">#&gt; [1] "Computing guide scores"</span>
<span class="co">#&gt; [1] "Blacklisting kmers"</span></code></pre></div>
</div>
<div id="compare-predicted-and-experimental-binding-sites" class="section level2">
<h2 class="hasAnchor">
<a href="#compare-predicted-and-experimental-binding-sites" class="anchor"></a>Compare predicted and experimental binding sites</h2>
<p>We can extract predicted binding sites using the <code>mappings</code> accessor and import provided dCAS9 ChIP peak coordinates from Fuentes et al. [1]. To facilitate comparison, we restrict our analysis to peaks on the main chromosome assemblies.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Extract predicted binding sites</span>
pred_sites &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mappings.html">mappings</a></span>(gs_fuentes)     

<span class="co"># Import experimental binding sites            </span>
peaks_path &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="dt">package =</span> <span class="st">'Repguide'</span>, <span class="st">'extdata'</span>, <span class="st">'fuentes_elife_2018_dCas9_peaks.bed'</span>)
exp_sites  &lt;-<span class="st"> </span>rtracklayer<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html">import.bed</a></span>(peaks_path)

<span class="co"># Throw chromosomes not present in guideSet</span>
<span class="kw">seqlevels</span>(exp_sites, <span class="dt">pruning.mode =</span> <span class="st">'coarse'</span>) &lt;-<span class="st"> </span><span class="kw">seqlevels</span>(pred_sites) </code></pre></div>
<p>We then compare the absolute number and binding intensities of correctly predicted binding sites.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hits &lt;-<span class="st"> </span>IRanges<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">findOverlaps</a></span>(exp_sites, pred_sites)

<span class="co"># Add logical whether or not peak was predicted by Repguide</span>
exp_sites<span class="op">$</span>predicted &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(exp_sites) <span class="op">%in%</span><span class="st"> </span>hits<span class="op">@</span>from

<span class="co"># Add Repguide predicted binding score to ChIP peaks</span>
exp_sites<span class="op">$</span>score_pred =<span class="st"> </span><span class="dv">0</span>
exp_sites[<span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(hits<span class="op">@</span>from)]<span class="op">$</span>score_pred &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(hits) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">score_pred =</span> pred_sites[hits<span class="op">@</span>to]<span class="op">$</span>Sbind) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(queryHits) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="kw">summarise</span>(<span class="dt">score_pred =</span> <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(score_pred)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">pull</span>(score_pred)</code></pre></div>
</div>
<div id="plotting-the-results" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-the-results" class="anchor"></a>Plotting the results</h2>
<p>We finally percent rank transform the predicted and experimental scores and visualize the results using <code>ggplot2</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ggdata &lt;-<span class="st"> </span>
<span class="st">  </span>exp_sites <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>as.data.frame <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">score_pred =</span> <span class="kw">percent_rank</span>(score_pred), 
         <span class="dt">score      =</span> <span class="kw">percent_rank</span>(score)) 

<span class="co"># Create plots         </span>
p_peaks_pie  &lt;-<span class="st"> </span>
<span class="st">  </span>ggdata <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> <span class="st">''</span>, <span class="dt">fill =</span> predicted)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_bar</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">coord_polar</span>(<span class="st">"y"</span>, <span class="dt">start =</span> <span class="dv">0</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">'none'</span>,
          <span class="dt">panel.border =</span> <span class="kw">element_blank</span>(),
          <span class="dt">axis.title.x =</span> <span class="kw">element_blank</span>(),
          <span class="dt">axis.title.y =</span> <span class="kw">element_blank</span>())
p_score_boxplot &lt;-<span class="st">              </span>
<span class="st">  </span>ggdata <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> predicted, <span class="dt">y =</span> score, <span class="dt">group =</span> predicted, <span class="dt">fill =</span> predicted)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_boxplot</span>(<span class="dt">outlier.shape =</span> <span class="ot">NA</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_jitter</span>(<span class="dt">size =</span> <span class="dv">1</span>, <span class="dt">alpha =</span> <span class="fl">0.1</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">90</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">'none'</span>)
p_score_scatter &lt;-<span class="st"> </span>
<span class="st">  </span>ggdata <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(predicted) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(score, score_pred)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">1</span>)

<span class="co"># Combine plots </span>
cowplot<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(p_peaks_pie, 
                   p_score_boxplot,
                   p_score_scatter, <span class="dt">rel_widths =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1.5</span>, <span class="dv">1</span>, <span class="dv">2</span>), 
                   <span class="dt">labels =</span> <span class="st">'auto'</span>, 
                   <span class="dt">nrow =</span> <span class="dv">1</span>)</code></pre></div>
<p><img src="benchmark_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We can see that Repguide correctly predicted more than 3/4 of experimentally determined Cas9 peaks (<strong>a</strong>) and failed predictions generally showed weak ChIP enrichment (<strong>b</strong>). Moreover, computed binding scores were largely consistent with ChIP signal intensities (<strong>c</strong>), suggesting Repguide on and off-targetd scores can be interpreted quantitatively.</p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>[1] Daniel R Fuentes, Tomek Swigut, Joanna Wysocka. Systematic perturbation of retroviral LTRs reveals widespread long-range effects on human gene regulation. eLife (2018).</p>
</div>
<div id="session-information" class="section level2">
<h2 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 3.5.3 (2019-03-11)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: CentOS Linux 7 (Core)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /net/mraid14/export/data/users/eladch/tools/CO7/mkl/2018.3/compilers_and_libraries_2018.3.222/linux/mkl/lib/intel64_lin/libmkl_gf_lp64.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8      </span>
<span class="co">#&gt;  [8] LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] forcats_0.4.0                     stringr_1.4.0                     dplyr_0.8.3                       purrr_0.3.2                       readr_1.3.1                      </span>
<span class="co">#&gt;  [6] tidyr_1.0.0                       tibble_2.1.3                      ggplot2_3.2.1                     tidyverse_1.2.1                   BSgenome.Hsapiens.UCSC.hg38_1.4.1</span>
<span class="co">#&gt; [11] BSgenome_1.50.0                   rtracklayer_1.42.2                Biostrings_2.50.2                 XVector_0.22.0                    GenomicRanges_1.34.0             </span>
<span class="co">#&gt; [16] GenomeInfoDb_1.18.2               IRanges_2.16.0                    S4Vectors_0.20.1                  BiocGenerics_0.28.0               Repguide_0.0.1.3                 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] nlme_3.1-141                bitops_1.0-6                matrixStats_0.55.0          fs_1.3.1                    doMC_1.3.6                  lubridate_1.7.4             httr_1.4.1                 </span>
<span class="co">#&gt;  [8] rprojroot_1.3-2             tools_3.5.3                 backports_1.1.4             R6_2.4.0                    lazyeval_0.2.2              colorspace_1.4-1            withr_2.1.2                </span>
<span class="co">#&gt; [15] tidyselect_0.2.5            compiler_3.5.3              cli_1.1.0                   rvest_0.3.4                 Biobase_2.42.0              xml2_1.2.2                  desc_1.2.0                 </span>
<span class="co">#&gt; [22] DelayedArray_0.8.0          labeling_0.3                scales_1.0.0                pkgdown_1.4.0               digest_0.6.20               Rsamtools_1.34.1            rmarkdown_1.15             </span>
<span class="co">#&gt; [29] R.utils_2.9.0               pkgconfig_2.0.2             htmltools_0.3.6             rlang_0.4.0                 readxl_1.3.1                rstudioapi_0.10             generics_0.0.2             </span>
<span class="co">#&gt; [36] jsonlite_1.6                BiocParallel_1.16.6         R.oo_1.22.0                 RCurl_1.95-4.12             magrittr_1.5                GenomeInfoDbData_1.2.0      Matrix_1.2-17              </span>
<span class="co">#&gt; [43] Rcpp_1.0.2                  munsell_0.5.0               Rbowtie_1.22.0              lifecycle_0.1.0             R.methodsS3_1.7.1           stringi_1.4.3               yaml_2.2.0                 </span>
<span class="co">#&gt; [50] MASS_7.3-51.4               SummarizedExperiment_1.12.0 zlibbioc_1.28.0             grid_3.5.3                  crayon_1.3.4                lattice_0.20-38             cowplot_1.0.0              </span>
<span class="co">#&gt; [57] haven_2.1.1                 hms_0.5.1                   zeallot_0.1.0               knitr_1.24                  pillar_1.4.2                codetools_0.2-16            XML_3.98-1.20              </span>
<span class="co">#&gt; [64] glue_1.3.1                  evaluate_0.14               data.table_1.12.2           modelr_0.1.5                vctrs_0.2.0                 foreach_1.4.7               cellranger_1.1.0           </span>
<span class="co">#&gt; [71] gtable_0.3.0                assertthat_0.2.1            xfun_0.9                    broom_0.5.2                 iterators_1.0.12            GenomicAlignments_1.18.1    memoise_1.1.0              </span>
<span class="co">#&gt; [78] ellipsis_0.2.0.1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#running-repguide">Running Repguide</a></li>
      <li><a href="#compare-predicted-and-experimental-binding-sites">Compare predicted and experimental binding sites</a></li>
      <li><a href="#plotting-the-results">Plotting the results</a></li>
      <li><a href="#references">References</a></li>
      <li><a href="#session-information">Session information</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://scholar.google.de/citations?user=ossdDj4AAAAJ&amp;hl=en&amp;oi=ao">David Brocks</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
